### PR TITLE
DBZ-1068 Forward port the work by @grzegorz8 to master

### PR DESCRIPTION
**Pull request by GuyIEX**
Co-authored-by: GuyIEX <gpascarella@ironeaglex.com>

### CODE REVIEW

**Comment by jpechane** [comment]
@GuyIEX Hi, could you please check the PR? It seems `EventProcessingFailureHandlingIT` is failing.

**Comment by GuyIEX** [comment]
@jpechane I was just working on that earlier today. I see why it's failing, but I'm trying to determine if those tests are there because this capability wasn't there before or if it's something I need to account for in this PR. I've updated my code locally to work around this, but it feels very hack-y. I basically check to see if an Error Handling mode is enabled and if it is then revert to the legacy behavior. If it's not then I use the other code path that allows column-based CDC.

Do you think it would be better to add a feature toggle for this or use the work-around I'm currently testing?

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTablePointer.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+     * @throws SQLException
+     */
+    protected boolean isNewTransaction() throws SQLException {
+        return (getPreviousChangePosition() != null) &&
+                getChangePosition().getCommitLsn().compareTo(getPreviousChangePosition().getCommitLsn()) > 0;
+    }
+
+    @Override
+    public Object[] getData() throws SQLException {
+        // final int dataColumnCount = resultSet.getMetaData().getColumnCount() - (columnDataOffset - 1);
```
Comment (jpechane) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTablePointer.java:90]: Please remove commented out code

**Comment by jpechane** [comment]
@GuyIEX I understand it now. The committed code no longer generates the error we rely on.
I'd then have a proposal to move forward

* copy current test and modify assertions to prove that this is no longer error - I think it is still good to keep test for this scenario
* modify  `EventProcessingFailureHandlingIT` with a different scenario that would generate error so we would still be able to confirm the auto restart works -maybe changing a column datatype?

WDYT?

**Comment by GuyIEX** [comment]
@jpechane That sounds like a great idea! I'll get to work on it.

**Comment by GuyIEX** [comment]
@jpechane Actually there was a problem in the code that I fixed to resolve the expected behavior in the tests. Things look good now for a merge. Thanks!

**Comment by gunnarmorling** [comment]
@jpechane, any more concerns with this one?

**Comment by jpechane** [comment]
@gunnarmorling From my POV this is good to go

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java [hunk 2]
```diff
@@ -360,56 +356,17 @@ private void migrateTable(final Queue<SqlServerChangeTable> schemaChangeCheckpoi
                                 dataConnection.getTableSchemaFromTable(currentTable),
                                 SchemaChangeEventType.CREATE));
             }
+            //
+            // TODO: This needs to be re-worked per https://github.com/debezium/debezium/pull/748#issuecomment-492526200
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java:360]: @jpechane, is this todo tracked in Jira?

Comment (gunnarmorling) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java:360]: @jpechane, ok, so I'm confused. This refers to the very comment that was the reason why @grzegorz8's PR wasn't merged in the first place. I just don't know: is this an actual issue now or not?

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTablePointer.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeTablePointer.java]: Pushed a commit for doing so.

**Comment by gunnarmorling** [comment]
@jpechane, please take a look at my comment inline. I'd *really* love to wrap up this one, I just don't know whether there is a remaining loose end here or not.

**Comment by gunnarmorling** [comment]
Rebased and applied. Logged DBZ-2495 for the remaining follow-up tasks. Thanks, @grzegorz8 and @GuyIEX!

**Pull request closed without merging**